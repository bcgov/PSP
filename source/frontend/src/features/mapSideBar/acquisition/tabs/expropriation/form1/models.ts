export class Form1Model {}
