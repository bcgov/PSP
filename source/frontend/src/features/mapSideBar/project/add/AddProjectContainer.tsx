import { FormikProps } from 'formik';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { FaBriefcase } from 'react-icons/fa';
import { useHistory } from 'react-router-dom';

import ConfirmNavigation from '@/components/common/ConfirmNavigation';
import * as API from '@/constants/API';
import MapSideBarLayout from '@/features/mapSideBar/layout/MapSideBarLayout';
import { useFinancialCodeRepository } from '@/hooks/repositories/useFinancialCodeRepository';
import useApiUserOverride from '@/hooks/useApiUserOverride';
import useLookupCodeHelpers from '@/hooks/useLookupCodeHelpers';
import { ApiGen_Concepts_FinancialCode } from '@/models/api/generated/ApiGen_Concepts_FinancialCode';
import { ApiGen_Concepts_FinancialCodeTypes } from '@/models/api/generated/ApiGen_Concepts_FinancialCodeTypes';
import { ApiGen_Concepts_Project } from '@/models/api/generated/ApiGen_Concepts_Project';
import { UserOverrideCode } from '@/models/api/UserOverrideCode';
import { toDropDownOptions } from '@/utils/financialCodeUtils';

import SidebarFooter from '../../shared/SidebarFooter';
import { useAddProjectForm } from '../hooks/useAddProjectFormManagement';
import { ProjectForm } from '../models';
import AddProjectForm from './AddProjectForm';

export interface IAddProjectContainerProps {
  onClose: () => void;
  onSuccess: (newProjectId: number) => void;
}

const AddProjectContainer: React.FC<React.PropsWithChildren<IAddProjectContainerProps>> = props => {
  const { onClose } = props;
  const history = useHistory();

  const {
    getFinancialCodesByType: { execute: getFinancialCodes },
  } = useFinancialCodeRepository();

  const [businessFunctions, setBusinessFunctions] = useState<ApiGen_Concepts_FinancialCode[]>([]);
  const [costTypes, setCostTypes] = useState<ApiGen_Concepts_FinancialCode[]>([]);
  const [workActivities, setWorkActivities] = useState<ApiGen_Concepts_FinancialCode[]>([]);
  const [isValid, setIsValid] = useState<boolean>(true);

  const withUserOverride = useApiUserOverride<
    (userOverrideCodes: UserOverrideCode[]) => Promise<ApiGen_Concepts_Project | void>
  >('Failed to Add Project File');

  useEffect(() => {
    async function fetchBusinessFunctions() {
      const data =
        (await getFinancialCodes(ApiGen_Concepts_FinancialCodeTypes.BusinessFunction)) ?? [];
      setBusinessFunctions(data);
    }
    fetchBusinessFunctions();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    async function fetchCostTypes() {
      const data = (await getFinancialCodes(ApiGen_Concepts_FinancialCodeTypes.CostType)) ?? [];
      setCostTypes(data);
    }
    fetchCostTypes();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    async function fetchWorkActivities() {
      const data = (await getFinancialCodes(ApiGen_Concepts_FinancialCodeTypes.WorkActivity)) ?? [];
      setWorkActivities(data);
    }
    fetchWorkActivities();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const businessFunctionOptions = useMemo(
    () => toDropDownOptions(businessFunctions),
    [businessFunctions],
  );

  const costTypeOptions = useMemo(() => toDropDownOptions(costTypes), [costTypes]);

  const workActivityOptions = useMemo(() => toDropDownOptions(workActivities), [workActivities]);

  const { getOptionsByType } = useLookupCodeHelpers();
  const projectStatusTypeCodes = getOptionsByType(API.PROJECT_STATUS_TYPES);

  const formikRef = useRef<FormikProps<ProjectForm>>(null);

  const close = useCallback(() => onClose(), [onClose]);

  const handleSave = async () => {
    const result = await (formikRef.current?.submitForm() ?? Promise.resolve());
    setIsValid(formikRef.current?.isValid ?? false);
    return result;
  };

  const onSuccess = async (proj: ApiGen_Concepts_Project) => {
    formikRef.current?.resetForm();
    props.onSuccess(proj.id);
  };

  const helper = useAddProjectForm({ onSuccess });

  const checkState = useCallback(() => {
    return formikRef?.current?.dirty && !formikRef?.current?.isSubmitting;
  }, [formikRef]);

  return (
    <MapSideBarLayout
      showCloseButton
      title="Create Project"
      icon={<FaBriefcase size={26} />}
      onClose={close}
      footer={
        <SidebarFooter onSave={handleSave} onCancel={close} displayRequiredFieldError={!isValid} />
      }
    >
      <AddProjectForm
        ref={formikRef}
        initialValues={helper.initialValues}
        projectStatusOptions={projectStatusTypeCodes}
        businessFunctionOptions={businessFunctionOptions ?? []}
        costTypeOptions={costTypeOptions ?? []}
        workActivityOptions={workActivityOptions ?? []}
        onSubmit={(projectForm, formikHelpers) =>
          withUserOverride((userOverrideCodes: UserOverrideCode[]) =>
            helper.handleSubmit(projectForm, formikHelpers, userOverrideCodes),
          )
        }
        validationSchema={helper.validationSchema}
        isCreating
      />
      <ConfirmNavigation navigate={history.push} shouldBlockNavigation={checkState} />
    </MapSideBarLayout>
  );
};

export default AddProjectContainer;
