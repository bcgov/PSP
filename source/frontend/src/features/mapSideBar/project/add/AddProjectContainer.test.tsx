import { Formik, FormikProps } from 'formik';
import { createMemoryHistory } from 'history';
import { forwardRef, RefObject } from 'react';

import { useFinancialCodeRepository } from '@/hooks/repositories/useFinancialCodeRepository';
import { useProjectProvider } from '@/hooks/repositories/useProjectProvider';
import { useUserInfoRepository } from '@/hooks/repositories/useUserInfoRepository';
import { mockLookups } from '@/mocks/lookups.mock';
import { mockProjectPostResponse } from '@/mocks/projects.mock';
import { getUserMock } from '@/mocks/user.mock';
import { getEmptyBaseAudit } from '@/models/defaultInitializers';
import { lookupCodesSlice } from '@/store/slices/lookupCodes';
import { toTypeCodeNullable } from '@/utils/formUtils';
import {
  act,
  createAxiosError,
  getMockRepositoryObj,
  render,
  RenderOptions,
  screen,
  userEvent,
} from '@/utils/test-utils';

import { ProjectForm } from '../models';
import AddProjectContainer, { IAddProjectContainerProps } from './AddProjectContainer';
import { IAddProjectFormProps } from './AddProjectForm';

const history = createMemoryHistory();

// Need to mock this library for unit tests
vi.mock('react-visibility-sensor', () => {
  return {
    default: vi.fn().mockImplementation(({ children }) => {
      if (children instanceof Function) {
        return children({ isVisible: true });
      }
      return children;
    }),
  };
});

const mockAddProjectApi = getMockRepositoryObj();

vi.mock('@/hooks/repositories/useProjectProvider');
vi.mocked(useProjectProvider, { partial: true }).mockReturnValue({
  addProject: mockAddProjectApi,
});

vi.mock('@/hooks/repositories/useFinancialCodeRepository');
vi.mocked(useFinancialCodeRepository, { partial: true }).mockReturnValue({
  getFinancialCodesByType: getMockRepositoryObj([]),
});

vi.mock('@/hooks/repositories/useUserInfoRepository');
vi.mocked(useUserInfoRepository).mockReturnValue({
  retrieveUserInfo: vi.fn(),
  retrieveUserInfoLoading: true,
  retrieveUserInfoResponse: {
    ...getUserMock(),
    userRegions: [
      {
        id: 1,
        userId: 5,
        regionCode: 1,
        region: toTypeCodeNullable(1),
        user: null,
        ...getEmptyBaseAudit(),
      },
      {
        id: 2,
        userId: 5,
        regionCode: 2,
        region: toTypeCodeNullable(2),
        user: null,
        ...getEmptyBaseAudit(),
      },
    ],
  },
});

let viewProps: IAddProjectFormProps | undefined;
let formikRef: RefObject<FormikProps<ProjectForm>> | undefined;

const TestView = forwardRef<FormikProps<ProjectForm>, IAddProjectFormProps>((props, ref) => {
  viewProps = props;
  formikRef = ref as any;

  return (
    <Formik<ProjectForm>
      enableReinitialize
      innerRef={ref}
      onSubmit={props.onSubmit}
      initialValues={props.initialValues}
    >
      {({ values }) => <>Content Rendered - {values.projectName}</>}
    </Formik>
  );
});

const onClose = vi.fn();
const onSuccess = vi.fn();

describe('AddProjectContainer component', () => {
  // render component under test
  const setup = async (
    renderOptions: RenderOptions & { props?: Partial<IAddProjectContainerProps> } = {},
  ) => {
    const utils = render(
      <AddProjectContainer
        onClose={renderOptions?.props?.onClose ?? onClose}
        onSuccess={renderOptions?.props?.onSuccess ?? onSuccess}
        View={TestView}
      />,
      {
        ...renderOptions,
        store: {
          [lookupCodesSlice.name]: { lookupCodes: mockLookups },
        },
        claims: [],
        history,
      },
    );

    // wait for effects to settle
    await act(async () => {});

    return {
      ...utils,
    };
  };

  beforeEach(() => {
    viewProps = undefined;
    formikRef = undefined;
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders as expected', async () => {
    const { asFragment } = await setup();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the underlying form', async () => {
    const { getByText } = await setup();
    expect(getByText(/Content Rendered/)).toBeVisible();
  });

  it('should save the form and navigate to details view when Save button is clicked', async () => {
    const formValues = new ProjectForm();
    formValues.projectName = 'TRANS-CANADA HWY - 10';
    formValues.projectNumber = '99999';
    formValues.region = 1;
    formValues.projectStatusType = 'AC';
    formValues.summary = 'NEW PROJECT';

    mockAddProjectApi.execute.mockResolvedValue(
      mockProjectPostResponse(
        1,
        2,
        'updated project name',
        'updated number',
        2,
        'AC',
        'updated summary',
      ),
    );

    await setup();

    expect(formikRef.current).not.toBeNull();
    await act(async () => formikRef.current?.setValues(formValues));
    await act(async () => formikRef.current?.submitForm());

    expect(mockAddProjectApi.execute).toHaveBeenCalledWith(formValues.toApi(), []);
    expect(onSuccess).toHaveBeenCalledWith(1);
  });

  it('displays server errors in a modal', async () => {
    mockAddProjectApi.execute.mockRejectedValue(
      createAxiosError(
        400,
        'Invalid Project management team, each team member can only be added once',
      ),
    );

    await setup();

    expect(formikRef.current).not.toBeNull();
    await act(async () => formikRef.current?.submitForm());

    expect(
      await screen.findByText(
        'Invalid Project management team, each team member can only be added once',
      ),
    ).toBeVisible();

    await act(async () => userEvent.click(screen.getByTitle('ok-modal')));
    expect(onSuccess).not.toHaveBeenCalled();
  });
});
