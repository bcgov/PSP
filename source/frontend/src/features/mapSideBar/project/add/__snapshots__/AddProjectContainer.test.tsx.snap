// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddProjectContainer component > renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  .c0 {
  font-family: 'BCSans-Bold';
  font-size: 3.2rem;
  border-bottom: solid 0.5rem;
  width: 100%;
  text-align: left;
  margin-bottom: 2.4rem;
}

.c2 {
  width: 100%;
  border-bottom: solid 0.5rem;
}

.c1 {
  font-size: 2.4rem;
  cursor: pointer;
}

<div
    class=""
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h1
          class="c0 mr-auto"
        >
          <svg
            class="mr-2 mb-2"
            fill="currentColor"
            height="32"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 336c0 8.84-7.16 16-16 16h-96c-8.84 0-16-7.16-16-16v-48H0v144c0 25.6 22.4 48 48 48h416c25.6 0 48-22.4 48-48V288H320v48zm144-208h-80V80c0-25.6-22.4-48-48-48H176c-25.6 0-48 22.4-48 48v48H48c-25.6 0-48 22.4-48 48v80h512v-80c0-25.6-22.4-48-48-48zm-144 0H192V96h128v32z"
            />
          </svg>
          Create Project
        </h1>
      </div>
      <div
        class="col-auto"
      >
        <svg
          class="c1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            close
          </title>
          <path
            d="M464 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-83.6 290.5c4.8 4.8 4.8 12.6 0 17.4l-40.5 40.5c-4.8 4.8-12.6 4.8-17.4 0L256 313.3l-66.5 67.1c-4.8 4.8-12.6 4.8-17.4 0l-40.5-40.5c-4.8-4.8-4.8-12.6 0-17.4l67.1-66.5-67.1-66.5c-4.8-4.8-4.8-12.6 0-17.4l40.5-40.5c4.8-4.8 12.6-4.8 17.4 0l66.5 67.1 66.5-67.1c4.8-4.8 12.6-4.8 17.4 0l40.5 40.5c4.8 4.8 4.8 12.6 0 17.4L313.3 256l67.1 66.5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="c2"
    />
  </div>
  .c10.btn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.4rem 1.2rem;
  min-height: 3rem;
  border: 0.2rem solid transparent;
  border-radius: 0.4rem;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.8rem;
  font-family: 'BCSans','Noto Sans',Verdana,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  cursor: pointer;
  height: 3.8rem;
}

.c10.btn:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  opacity: 0.8;
}

.c10.btn:focus {
  outline-width: 0.4rem;
  outline-style: solid;
  outline-offset: 1px;
  box-shadow: none;
}

.c10.btn.btn-primary {
  border: none;
}

.c10.btn.btn-secondary {
  background: none;
}

.c10.btn.btn-info {
  border: none;
  background: none;
  padding-left: 0.6rem;
  padding-right: 0.6rem;
}

.c10.btn.btn-info:hover,
.c10.btn.btn-info:active,
.c10.btn.btn-info:focus {
  background: none;
}

.c10.btn.btn-light {
  border: none;
}

.c10.btn.btn-dark {
  border: none;
}

.c10.btn.btn-link {
  font-size: 1.6rem;
  font-weight: 400;
  background: none;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 2.5rem;
  line-height: 3rem;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  -webkit-letter-spacing: unset;
  -moz-letter-spacing: unset;
  -ms-letter-spacing: unset;
  letter-spacing: unset;
  text-align: left;
  padding: 0;
}

.c10.btn.btn-link:hover,
.c10.btn.btn-link:active,
.c10.btn.btn-link:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  border: none;
  background: none;
  box-shadow: none;
  outline: none;
}

.c10.btn.btn-link:disabled,
.c10.btn.btn-link.disabled {
  background: none;
  pointer-events: none;
}

.c10.btn:disabled,
.c10.btn:disabled:hover {
  box-shadow: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  cursor: not-allowed;
  opacity: 0.65;
}

.c10.Button .Button__icon {
  margin-right: 1.6rem;
}

.c10.Button--icon-only:focus {
  outline: none;
}

.c10.Button--icon-only .Button__icon {
  margin-right: 0;
}

.c9 button.close {
  font-size: 2.4rem;
}

.c0 {
  width: 100%;
  position: relative;
  overflow: auto;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  height: 100%;
  width: 100%;
}

.c4 {
  font-weight: bold;
  border-bottom: 0.2rem solid;
  margin-bottom: 2rem;
}

.c3 {
  margin: 1.6rem;
  padding: 1.6rem;
  background-color: white;
  text-align: left;
  border-radius: 0.5rem;
}

.c7.required::before {
  content: '*';
  position: absolute;
  top: 0.75rem;
  left: 0rem;
}

.c6 {
  font-weight: bold;
}

.c2 {
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}

.c5 {
  border-bottom-style: solid;
  border-bottom-color: grey;
  border-bottom-width: 0.1rem;
  text-align: left;
}

.c8 textarea.form-control {
  min-width: 80rem;
  height: 7rem;
  resize: none;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <form
          class=""
          novalidate=""
        >
          <div
            class="c3 form-section"
          >
            <h2
              class="c4"
            >
              <div
                class="no-gutters row"
              >
                <div
                  class="col"
                >
                  Project Details
                </div>
              </div>
            </h2>
            <div
              class="collapse show"
            >
              <div
                class="c5 no-gutters py-4 mb-5 row"
              >
                <div
                  class="col"
                >
                  <p>
                    Before creating a project, 
                    <a
                      href="/project/list"
                    >
                      do a search
                    </a>
                     to ensure the the project you're creating doesn't already exist.
                  </p>
                </div>
              </div>
              <div
                class="pb-2 row"
              >
                <div
                  class="pr-0 text-left col-2"
                >
                  <label
                    class="c6"
                  >
                    Project name:
                  </label>
                </div>
                <div
                  class="c7 required text-left col"
                >
                  <div
                    class="input form-group"
                  >
                    <input
                      class="form-control"
                      id="input-projectName"
                      name="projectName"
                      title=""
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="pb-2 row"
              >
                <div
                  class="pr-0 text-left col-2"
                >
                  <label
                    class="c6"
                  >
                    Project number:
                  </label>
                </div>
                <div
                  class="c7 text-left col"
                >
                  <div
                    class="input form-group"
                  >
                    <input
                      class="form-control"
                      id="input-projectNumber"
                      name="projectNumber"
                      placeholder="if known"
                      title=""
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="pb-2 row"
              >
                <div
                  class="pr-0 text-left col-2"
                >
                  <label
                    class="c6"
                  >
                    Status:
                  </label>
                </div>
                <div
                  class="c7 required text-left col"
                >
                  <div
                    class="form-group"
                  >
                    <select
                      class="form-select form-control"
                      id="input-projectStatusType"
                      name="projectStatusType"
                    >
                      <option
                        value=""
                      >
                        Select...
                      </option>
                      <option
                        class="option"
                        data-testid="select-option-AC"
                        value="AC"
                      >
                        Active (AC)
                      </option>
                      <option
                        class="option"
                        data-testid="select-option-CA"
                        value="CA"
                      >
                        Cancelled (CA)
                      </option>
                      <option
                        class="option"
                        data-testid="select-option-CNCN"
                        value="CNCN"
                      >
                        Consolidated (CNCN)
                      </option>
                      <option
                        class="option"
                        data-testid="select-option-CO"
                        value="CO"
                      >
                        Completed (CO)
                      </option>
                      <option
                        class="option"
                        data-testid="select-option-HO"
                        value="HO"
                      >
                        On Hold (HO)
                      </option>
                      <option
                        class="option"
                        data-testid="select-option-PL"
                        value="PL"
                      >
                        Planning (PL)
                      </option>
                    </select>
                  </div>
                </div>
              </div>
              <div
                class="pb-2 row"
              >
                <div
                  class="pr-0 text-left col-2"
                >
                  <label
                    class="c6"
                  >
                    MoTI region:
                  </label>
                </div>
                <div
                  class="c7 required text-left col"
                >
                  <div
                    class="d-flex form-group"
                  >
                    <select
                      class="form-select form-control"
                      id="input-region"
                      name="region"
                    >
                      <option
                        value=""
                      >
                        Select region...
                      </option>
                      <option
                        class="option"
                        data-testid="select-option-1"
                        value="1"
                      >
                        South Coast Region
                      </option>
                      <option
                        class="option"
                        data-testid="select-option-2"
                        value="2"
                      >
                        Southern Interior Region
                      </option>
                    </select>
                  </div>
                </div>
              </div>
              <div
                class="pb-2 row"
              >
                <div
                  class="pr-0 text-left col-12"
                >
                  <label
                    class="c6"
                  >
                    Project summary:
                  </label>
                </div>
                <div
                  class="c7 text-left col"
                >
                  <div
                    class="c8 textarea input form-group"
                  >
                    <textarea
                      class="form-control"
                      id="input-summary"
                      name="summary"
                      title=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c3 form-section"
          >
            <h2
              class="c4"
            >
              <div
                class="no-gutters row"
              >
                <div
                  class="col"
                >
                  Associated Codes
                </div>
              </div>
            </h2>
            <div
              class="collapse show"
            >
              <div
                class="pb-2 row"
              >
                <div
                  class="pr-0 text-left col-2"
                >
                  <label
                    class="c6"
                  >
                    Cost type:
                  </label>
                </div>
                <div
                  class="c7 text-left col"
                >
                  <div
                    class="c9 form-group"
                    data-testid="typeahead-select-costTypeCode"
                  >
                    <div
                      class="rbt"
                      style="outline: none; position: relative;"
                      tabindex="-1"
                    >
                      <div
                        style="display: flex; flex: 1; height: 100%; position: relative;"
                      >
                        <input
                          aria-autocomplete="both"
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          autocomplete="off"
                          class="rbt-input-main form-control rbt-input"
                          id="typeahead-select-costTypeCode"
                          name="typeahead-select-costTypeCode"
                          placeholder="Type to search..."
                          role="combobox"
                          type="text"
                          value=""
                        />
                        <input
                          aria-hidden="true"
                          class="rbt-input-hint"
                          readonly=""
                          style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.54); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%; border-style: inset inset inset inset; border-width: 2px 2px 2px 2px; line-height: normal; padding: 1px 1px 1px 1px;"
                          tabindex="-1"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="pb-2 row"
              >
                <div
                  class="pr-0 text-left col-2"
                >
                  <label
                    class="c6"
                  >
                    Work activity:
                  </label>
                </div>
                <div
                  class="c7 text-left col"
                >
                  <div
                    class="c9 form-group"
                    data-testid="typeahead-select-workActivityCode"
                  >
                    <div
                      class="rbt"
                      style="outline: none; position: relative;"
                      tabindex="-1"
                    >
                      <div
                        style="display: flex; flex: 1; height: 100%; position: relative;"
                      >
                        <input
                          aria-autocomplete="both"
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          autocomplete="off"
                          class="rbt-input-main form-control rbt-input"
                          id="typeahead-select-workActivityCode"
                          name="typeahead-select-workActivityCode"
                          placeholder="Type to search..."
                          role="combobox"
                          type="text"
                          value=""
                        />
                        <input
                          aria-hidden="true"
                          class="rbt-input-hint"
                          readonly=""
                          style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.54); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%; border-style: inset inset inset inset; border-width: 2px 2px 2px 2px; line-height: normal; padding: 1px 1px 1px 1px;"
                          tabindex="-1"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="pb-2 row"
              >
                <div
                  class="pr-0 text-left col-2"
                >
                  <label
                    class="c6"
                  >
                    Business function:
                  </label>
                </div>
                <div
                  class="c7 text-left col"
                >
                  <div
                    class="c9 form-group"
                    data-testid="typeahead-select-businessFunctionCode"
                  >
                    <div
                      class="rbt"
                      style="outline: none; position: relative;"
                      tabindex="-1"
                    >
                      <div
                        style="display: flex; flex: 1; height: 100%; position: relative;"
                      >
                        <input
                          aria-autocomplete="both"
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          autocomplete="off"
                          class="rbt-input-main form-control rbt-input"
                          id="typeahead-select-businessFunctionCode"
                          name="typeahead-select-businessFunctionCode"
                          placeholder="Type to search..."
                          role="combobox"
                          type="text"
                          value=""
                        />
                        <input
                          aria-hidden="true"
                          class="rbt-input-hint"
                          readonly=""
                          style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.54); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%; border-style: inset inset inset inset; border-width: 2px 2px 2px 2px; line-height: normal; padding: 1px 1px 1px 1px;"
                          tabindex="-1"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c3 form-section"
          >
            <h2
              class="c4"
            >
              <div
                class="no-gutters row"
              >
                <div
                  class="col"
                >
                  Associated products
                </div>
              </div>
            </h2>
            <div
              class="collapse show"
            >
              <button
                class="c10 Button btn btn-link"
                type="button"
              >
                <div
                  class="Button__value"
                >
                  + Add another product
                </div>
              </button>
            </div>
          </div>
          <div />
          <div />
        </form>
      </div>
    </div>
  </div>
  .c1.btn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.4rem 1.2rem;
  min-height: 3rem;
  border: 0.2rem solid transparent;
  border-radius: 0.4rem;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.8rem;
  font-family: 'BCSans','Noto Sans',Verdana,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  cursor: pointer;
  height: 3.8rem;
}

.c1.btn:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  opacity: 0.8;
}

.c1.btn:focus {
  outline-width: 0.4rem;
  outline-style: solid;
  outline-offset: 1px;
  box-shadow: none;
}

.c1.btn.btn-primary {
  border: none;
}

.c1.btn.btn-secondary {
  background: none;
}

.c1.btn.btn-info {
  border: none;
  background: none;
  padding-left: 0.6rem;
  padding-right: 0.6rem;
}

.c1.btn.btn-info:hover,
.c1.btn.btn-info:active,
.c1.btn.btn-info:focus {
  background: none;
}

.c1.btn.btn-light {
  border: none;
}

.c1.btn.btn-dark {
  border: none;
}

.c1.btn.btn-link {
  font-size: 1.6rem;
  font-weight: 400;
  background: none;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 2.5rem;
  line-height: 3rem;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  -webkit-letter-spacing: unset;
  -moz-letter-spacing: unset;
  -ms-letter-spacing: unset;
  letter-spacing: unset;
  text-align: left;
  padding: 0;
}

.c1.btn.btn-link:hover,
.c1.btn.btn-link:active,
.c1.btn.btn-link:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  border: none;
  background: none;
  box-shadow: none;
  outline: none;
}

.c1.btn.btn-link:disabled,
.c1.btn.btn-link.disabled {
  background: none;
  pointer-events: none;
}

.c1.btn:disabled,
.c1.btn:disabled:hover {
  box-shadow: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  cursor: not-allowed;
  opacity: 0.65;
}

.c1.Button .Button__icon {
  margin-right: 1.6rem;
}

.c1.Button--icon-only:focus {
  outline: none;
}

.c1.Button--icon-only .Button__icon {
  margin-right: 0;
}

.c0 {
  position: -webkit-sticky;
  position: sticky;
  padding-top: 2rem;
  bottom: 0;
  background: white;
  z-index: 10;
}

<div
    class=""
  >
    <div
      class="c0 justify-content-end mb-12 mt-12 p-0 no-gutters row"
    >
      <div
        class="pr-3 col-auto"
      />
      <div
        class="pr-6 col-auto"
      >
        <button
          class="c1 Button btn btn-secondary"
          type="button"
        >
          <div
            class="Button__value"
          >
            Cancel
          </div>
        </button>
      </div>
      <div
        class="col-auto"
      >
        <button
          class="c1 Button mr-9 btn btn-primary"
          type="button"
        >
          <div
            class="Button__value"
          >
            Save
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
